pkg/operators/ebpf/ebpf.go

```diff
@@ -0,0 +1,549 @@
+		}
+		m.ds = ds
+	}
+	for _, m := range i.snapshotters {
+		ds, err := gadgetCtx.RegisterDataSource(datasource.TypeEvent, m.StructName)
+		if err != nil {
+			return fmt.Errorf("adding tracer datasource: %w", err)
+		}
+		fields := make([]datasource.Field, 0, len(i.structs[m.StructName].Fields))
+		for _, field := range i.structs[m.StructName].Fields {
+			fields = append(fields, field)
+		}
+		m.accessor, err = ds.AddStaticFields(i.structs[m.StructName].Size, fields)
+		if err != nil {
+			return fmt.Errorf("adding fields for datasource: %w", err)
+		}
+		m.ds = ds
+	}
+	return nil
+}
+
+func (i *ebpfInstance) Name() string {
+	return "ebpfInstance"
+}
+
+func (i *ebpfInstance) GadgetParams() []*api.Param {
+	res := make([]*api.Param, 0, len(i.params))
+	for _, p := range i.params {
+		res = append(res, p)
+	}
+	return res
+}
+
+func (i *ebpfInstance) Prepare(gadgetCtx operators.GadgetContext) error {
+	for ds, converters := range i.converters {
+		for _, converter := range converters {
+			converter := converter
+			ds.Subscribe(func(ds datasource.DataSource, data datasource.Data) error {
+				return converter(ds, data)
+
+			}, 0)
+		}
+	}
+
+	// Create network tracers, one for each socket filter program
+	for _, p := range i.collectionSpec.Programs {
+		switch p.Type {
+		case ebpf.SocketFilter:
+			if strings.HasPrefix(p.SectionName, "socket") {
+				networkTracer, err := networktracer.NewTracer[api.GadgetData]()
+				if err != nil {
+					i.Close()
+					return fmt.Errorf("creating network tracer: %w", err)
+				}
+				i.networkTracers[p.Name] = networkTracer
+			}
+		case ebpf.SchedCLS:
+			parts := strings.Split(p.SectionName, "/")
+			if len(parts) != 3 {
+				return fmt.Errorf("invalid section name %q", p.SectionName)
+			}
+			if parts[0] != "classifier" {
+				return fmt.Errorf("invalid section name %q", p.SectionName)
+			}
+
+			var direction tchandler.AttachmentDirection
+
+			switch parts[1] {
+			case "ingress":
+				direction = tchandler.AttachmentDirectionIngress
+			case "egress":
+				direction = tchandler.AttachmentDirectionEgress
+			default:
+				return fmt.Errorf("unsupported hook type %q", parts[1])
+			}
+
+			handler, err := tchandler.NewHandler(direction)
+			if err != nil {
+				i.Close()
+				return fmt.Errorf("creating tc network tracer: %w", err)
+			}
+
+			i.tcHandlers[p.Name] = handler
+		}
+	}
+	return nil
+}
+
+func (i *ebpfInstance) Start(gadgetCtx operators.GadgetContext) error {
+	i.logger.Debugf("starting ebpfInstance")
+
+	gadgets.FixBpfKtimeGetBootNs(i.collectionSpec.Programs)
+
+	parms := params.Params{}                   // used to CopyFromMap
+	paramMap := make(map[string]*params.Param) // used for second iteration
+	for name, p := range i.params {
+		param := api.ParamToParamDesc(p).ToParam()
+		paramMap[name] = param
+		parms = append(parms, param)
+	}
+	err := parms.CopyFromMap(i.paramValues, "")
+	if err != nil {
+		return fmt.Errorf("parsing parameter values: %w", err)
+	}
```

Comment:
I've tried to make it work with the params package but failed at one point or another. Params are tricky because we add them at different stages of a gadgets' lifecycle, some of them have to be available early on, some of them get their default values only in later stages - and it all needs to be serializable to make it work remotely. This right now feels like a compromise that is just _good enough_:

While we always serialize/deserialize into/from maps, even when used locally, this format is the same that we will most likely use for the GadgetInstanceManager (gadget configuration files) and that gets sent over the wire. Both, the local runtime and the gRPC runtime can understand it well and at any given point in the lifecycle, components requiring the param values can extract them.

I'll keep thinking about alternatives and when all components are merged (GadgetInstanceManager, external interfaces), but in this PR, for now, I would like to go ahead with it.